### PR TITLE
fix(1242): add scenario of branch filtering

### PR DESCRIPTION
**Pull request by tk3fftk**
## Context
There is not feature test for branch filtering workflow.

## Objective
This PR adds scenarios for branch filtering workflow.
Also updates pipeline definition of other scenarios to using `requires`.

## References
design doc: https://github.com/screwdriver-cd/screwdriver/blob/ba6478f5d43acae3fa1d5396cc2c417a591cee7a/design/workflow.md
guide: https://docs.screwdriver.cd/user-guide/configuration/workflow#branch-filtering
related issue: https://github.com/screwdriver-cd/screwdriver/issues/1242

### CODE REVIEW

File: features/workflow.feature [hunk 1]
```diff
@@ -58,3 +64,55 @@ Feature: Workflow
...
+            | FOO   | ~pr             |
+            | BAR   | ~pr:staging     |
+            | BAZ   | ~pr:/^.*$/      |
+            | QUX   | ~pr:/master     |
+        When a pull request is opened
+        And it is targeting the pipeline's branch
+        Then the "FOO" job is started
+        And the "BAR" job is not started
+        And the "BAZ" job is started
+        And the "QUX" job is started
```
Comment (tk3fftk) [inline features/workflow.feature]: "QUX" job is not started in current implementation.
ref: https://github.com/screwdriver-cd/screwdriver/blob/ba6478f5d43acae3fa1d5396cc2c417a591cee7a/plugins/webhooks/index.js#L176

Do we need to allow and work `~pr:<pipeline_branch>` ?

**Comment by yuichi10** [comment]
I have questions about the specific PR workflow.
1) Should specific PR run when the specific branch is same with the pipeline branch? (Same with https://github.com/screwdriver-cd/screwdriver/pull/1286#discussion_r216532684 )
2) Should specific PR allow to trigger?
3) Should specific PR allow to publish or deploy?

I want to know how the spcific PR workflow work.

**Comment by tk3fftk** [comment]
In my opinion, 1-3 should be worked. This feature will fix https://github.com/screwdriver-cd/screwdriver/issues/1085 by another implementation.

**Comment by yuichi10** [comment]
I see. I also think it is a good solution that specific PR allow 1-3 for the issue!
Then do we need the test about specific PR triggers? Also, do we need to test that ~PR will not trigger?

**Comment by tk3fftk** [comment]
> Then do we need the test about specific PR triggers? Also, do we need to test that ~PR will not trigger?

I'll add trigger test cases.

File: features/workflow.feature [hunk 2]
```diff
@@ -58,3 +64,71 @@ Feature: Workflow
...
+        Given an existing pipeline with the workflow:
+            | job   | requires        |
+            | FOO   | ~commit         |
+            | BAR   | ~commit:staging |
+            | BAZ   | ~commit:/^.*$/  |
+        When a new commit is pushed
+        And it is against the pipeline's branch
+        Then the "FOO" job is started
+        And the "BAR" job is not started
+        And the "BAZ" job is started
```
[the remark this answers is not in the document]
Comment (tk3fftk) [reply features/workflow.feature]: Other cases check SHA with triggered job. This case a job is triggered by commit so I think it does not need in this case. But other scenario may need SHA check. Thanks.

File: features/workflow.feature [hunk 3]
```diff
@@ -58,3 +64,71 @@ Feature: Workflow
...
+            | FOO2  | FOO             |
+            | QUX   | ~pr:/master     |
+            | QUX2  | QUX             |
+        When a pull request is opened
+        And it is targeting the pipeline's branch
+        Then the "FOO" job is started
+        And the "QUX" job is started
+        Then the "FOO" job is succeeded
+        And the "QUX" job is succeeded
+        Then the "FOO2" job is not started
```
Comment (yoshwata) [inline features/workflow.feature]: should FOO2 be started?

Comment (tk3fftk) [reply features/workflow.feature]: This example is a little strange but `FOO2` is not started. `FOO2` job requires `FOO` and `FOO` requires `~pr`.  `~pr` will not trigger subsequence jobs.

File: features/workflow.feature [hunk 4]
```diff
@@ -46,15 +50,91 @@ Feature: Workflow
...
+        And that "QUX" build uses the same SHA as the "FOO" build
+
+    Scenario: Branch filtering (the pipeline's branch is committed)
+        Given an existing pipeline with the workflow:
+            | job   | requires        |
+            | FOO   | ~commit         |
+            | BAR   | ~commit:staging |
+            | BAZ   | ~commit:/^.*$/  |
+        When a new commit is pushed
+        And it is against the pipeline's branch
```
Comment (tkyi) [inline features/workflow.feature]: I think `pipeline's branch` wording is confusing for me
Can we call out the branch specifically by saying something like `And it is on the "master" branch`

And is there a difference between `it is against` vs `it is targeting`?

File: features/workflow.feature [hunk 5]
```diff
@@ -18,25 +18,29 @@ Feature: Workflow
...
         When the "FOO" job is started
         And the build succeeded
         Then the "BAR" job is started
         And that "BAR" build uses the same SHA as the "FOO" build
 
     Scenario: Parallel
         Given an existing pipeline with the workflow:
-            | job | triggers |
-            | FOO | BAR+BAZ  |
+            | job | requires |
```
Comment (tkyi) [inline features/workflow.feature]: 👍 thumbs up to changing from `job/triggers` to `job/requires`

File: features/workflow.feature (hunk 4 above)
Comment (tk3fftk) [reply features/workflow.feature]: `pipeline's branch` is reused from git-flow feature:
https://github.com/screwdriver-cd/screwdriver/blob/a37c2485340e84240bef8147d2f9bdd337c6f082/features/git-flow.feature#L37
But I agree with you that it's confusing so I fixed.

> And is there a difference between it is against vs it is targeting?

unified with `it is targeting`

Comment (tk3fftk) [reply features/workflow.feature]: @tkyi fixed!

**Pull request merged**